core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala

```diff
@@ -0,0 +1,320 @@
+    }
+
+
+    if (failureStatus.totalFailures >= MAX_FAILURES_PER_EXEC_STAGE) {
+      // this executor has been pushed into the blacklist for this stage.  Lets check if it pushes
+      // the whole node into the blacklist
+      val blacklistedExecutors =
+        stageFailures.filter{_._2.totalFailures >= MAX_FAILURES_PER_EXEC_STAGE}
+      if (blacklistedExecutors.size >= MAX_FAILED_EXEC_PER_NODE_STAGE) {
+        logInfo(s"Blacklisting ${info.host} for stage $stageId")
+        stageIdToBlacklistedNodes.getOrElseUpdate(stageId, new HashSet()) += info.host
+      }
+    }
+  }
+
+  /**
+   * Return true if this executor is blacklisted for the given task.  This does *not*
+   * need to return true if the executor is blacklisted for the entire stage, or blacklisted
+   * altogether.
+   */
+  def isExecutorBlacklisted(
+      executorId: String,
+      stageId: Int,
+      indexInTaskSet: Int): Boolean = {
+    // intentionally avoiding .getOrElse(..., new HashMap()) to avoid lots of object
+    // creation, since this method gets called a *lot*
+    stageIdToExecToFailures.get(stageId) match {
+      case Some(stageFailures) =>
+        stageFailures.get(executorId) match {
+          case Some(failures) =>
+            failures.failuresByTask.contains(indexInTaskSet)
+          case None =>
+            false
+        }
+      case None =>
+        false
+    }
+  }
+
+  def isNodeBlacklisted(
+      node: String,
+      stageId: Int,
+      indexInTaskSet: Int): Boolean = {
+    stageIdToNodeBlacklistedTasks.get(stageId).flatMap { nodeToFailures =>
+      nodeToFailures.get(node).map{_.contains(indexInTaskSet)}
+    }.getOrElse(false)
+  }
+
+  def removeExecutor(executorId: String): Unit = {
+    executorIdToBlacklistExpiryTime -= executorId
+    executorIdToFailureCount -= executorId
+    stageIdToExecToFailures.values.foreach { execFailureOneStage =>
+      execFailureOneStage -= executorId
+    }
+  }
+}
+
+
+private[scheduler] object BlacklistTracker extends Logging {
+  val LEGACY_TIMEOUT_CONF = "spark.scheduler.executorTaskBlacklistTime"
+  val EXPIRY_TIMEOUT_CONF = "spark.scheduler.blacklist.recoverPeriod"
+  val ENABLED_CONF = "spark.scheduler.blacklist.enabled"
+
+  /**
+   * Return true if the blacklist is enabled, based on the following order of preferences:
+   * 1. Is it specifically enabled or disabled?
+   * 2. Is it enabled via the legacy timeout conf?
+   * 3. Use the default for the spark-master:
+   *   - off for local mode
+   *   - on for distributed modes (including local-cluster)
+   */
+  def isBlacklistEnabled(conf: SparkConf): Boolean = {
+    val isEnabled = conf.get(ENABLED_CONF, null)
+    if (isEnabled == null) {
+      // if they've got a non-zero setting for the legacy conf, always enable the blacklist,
+      // otherwise, use the default based on the cluster-mode (off for local-mode, on otherwise).
+      val legacyTimeout = conf.getLong(LEGACY_TIMEOUT_CONF, 0L)
```

Comment:
If I understand this correctly, if a user has explicitly set `spark.scheduler.executorTaskBlacklistTime=0` in the configuration, that will fail to disable the blacklist, right?

Maybe you should be using `getOption` here to account for that case?